### PR TITLE
[Transform] add support for filter aggregation

### LINKED ISSUE #52151: [Transform] add support for filter aggregation
[Filter aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-filter-aggregation.html#) cover some nice usecases.

For example to gather stats for response codes:
```
   "aggregations": {
      "404": {
        "filter": {
          "term": {
            "response": "404"
          }
        }
      },
      "200": {
        "filter": {
          "term": {
            "response": "200"
          }
        }
      },
      "503": {
        "filter": {
          "term": {
            "response": "503"
          }
        }
      }
```

Note that `filter` supports sub aggregations which makes it hard to decide on the right structure in the transform destination index. For the simple example above the output structure could be:

```
   {
      "geo" : {
        "src" : "CM"
      },
      "200" : 41,
      "404" : 2,
      "503" : 0
    },
    {
      "geo" : {
        "src" : "CN"
      },
      "200" : 2415,
      "404" : 138,
      "503" : 89
    },
    {
      "geo" : {
        "src" : "CO"
      },
      "200" : 76,
      "404" : 8,
      "503" : 3
    },
```

For this the `doc_count` of the ouput is used as _flat_ result.

If you specify a sub-aggregation, we can not provide a flat result, because we need a nested object. For this case we could fallback to:

```
"my_agg_field": {
    "doc_count": 42,
    "sub_agg_field": {
        # sub agg result  
    }  
}
```

If a sub aggregation is used, the user probably does not care about the `doc_count` field, but we do not know. For getting rid of id, you can use a pipeline.

## Discuss

 - Should we have flattened results if `filter` specifies no sub aggregation?
 - If a sub-agg is given, how should the result look like?

### PR DESCRIPTION
**Pull request by hendrikmuhs**
add support for filter aggregations, refactor code for sub-aggregation support in mapping deduction

fixes #52151

**Note:** Dear reviewer, please start with https://github.com/elastic/elasticsearch/commit/78b63e334cd3c10cfcf3b399048cd453d22734ef

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java [hunk 1]
```diff
@@ -211,17 +215,35 @@ public Object value(Aggregation agg, String fieldType) {
...
+        public Object value(Aggregation agg, String fieldType, Map<String, String> fieldTypeMap) {
+            SingleBucketAggregation aggregation = (SingleBucketAggregation) agg;
+
+            if (aggregation.getAggregations().iterator().hasNext() == false) {
+                return aggregation.getDocCount();
+            }
+
+            HashMap<String, Object> nested = new HashMap<>();
+            for (Aggregation subAgg : aggregation.getAggregations()) {
+                nested.put(subAgg.getName(), getExtractor(subAgg).value(subAgg, fieldTypeMap.get(agg.getName()), fieldTypeMap));
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java]:
```suggestion
                nested.put(subAgg.getName(), getExtractor(subAgg).value(subAgg, fieldTypeMap.get(subAgg.getName()), fieldTypeMap));
```

We should use the `subAgg` field type no? If we just used the parent agg's type, we could pass `fieldType`

Comment (benwtrent) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java]: 🤔 

Does this actually work? It seems like we should recursively supply the `String fieldName` or something and then read that from `fieldTypeMap`. 

That way we can continually append the field name. 

Something like
```
nested.put(subAgg.getName(), getExtractor(subAgg).value(subAgg, fieldName + "." + subAgg.getName(), fieldTypeMap));
```

And those that worry about `fieldType` can get it from the `fieldTypeMap`

**Review by benwtrent — COMMENTED** [review]
I really like how this recursive parsing and dot delimited naming of fields work so well :).

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java (hunk 1 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java]: good catch! I think that makes sense.

In order to fix this, I will create a test combines both mapping deduction and value extraction, both are tested individually at the moment which makes it hard to discover bugs like this.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java [hunk 2]
```diff
@@ -211,17 +215,44 @@ public Object value(Aggregation agg, String fieldType) {
...
+            }
+
+            HashMap<String, Object> nested = new HashMap<>();
+            for (Aggregation subAgg : aggregation.getAggregations()) {
+                String subLookupFieldPrefix = lookupFieldPrefix.isEmpty() ? agg.getName() : lookupFieldPrefix + "." + agg.getName();
+                nested.put(
+                    subAgg.getName(),
+                    getExtractor(subAgg).value(
+                        subAgg,
+                        fieldTypeMap.get(subLookupFieldPrefix + "." + subAgg.getName()),
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java]: It strikes me that supplying `fieldType` is no longer necessary. 

Anything that desires to know `fieldType` could look it up via its own agg name combined with the prefix (if not empty).

**Review by benwtrent — APPROVED** [review]
minor observation. Otherwise, looks good :). Super excited to have these in transforms!

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc 
run elasticsearch-ci/default-distro

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/bwc

**Pull request merged**